src/objects/BatchedMesh.ts

```diff
@@ -0,0 +1,425 @@
+
+    return this
+  }
+
+  optimize() {
+    // @TODO: Implement
+
+    return this
+  }
+
+  setMatrixAt(geometryId: number, matrix: Matrix4) {
+    // @TODO: Map geometryId to index of the arrays because
+    //        optimize() can make geometryId mismatch the index
+
+    if (geometryId >= this._matrices.length || this._alives[geometryId] === false) {
+      return this
+    }
+
+    this._matrices[geometryId].copy(matrix)
+
+    if (this._visibles[geometryId] === true) {
+      matrix.toArray(this._matricesArray!, geometryId * 16)
+      this._matricesTexture!.needsUpdate = true
+    }
+
+    return this
+  }
+
+  getMatrixAt(geometryId: number, matrix: Matrix4) {
+    if (geometryId >= this._matrices.length || this._alives[geometryId] === false) {
+      return matrix
+    }
+
+    return matrix.copy(this._matrices[geometryId])
+  }
+
+  setVisibleAt(geometryId: number, visible: boolean) {
+    if (geometryId >= this._visibles.length || this._alives[geometryId] === false) {
+      return this
+    }
+
+    if (this._visibles[geometryId] === visible) {
+      return this
+    }
+
+    if (visible === true) {
+      this._matrices[geometryId].toArray(this._matricesArray!, geometryId * 16)
+    } else {
+      _zeroMatrix.toArray(this._matricesArray!, geometryId * 16)
+    }
+
+    this._matricesTexture!.needsUpdate = true
+    this._visibles[geometryId] = visible
+    return this
+  }
+
+  getVisibleAt(geometryId: number) {
+    if (geometryId >= this._visibles.length || this._alives[geometryId] === false) {
+      return false
+    }
+
+    return this._visibles[geometryId]
+  }
+
+  copy(source: BatchedMesh) {
+    // @ts-ignore
+    super.copy(source)
+
+    // @TODO: Implement
+
+    return this
+  }
+
+  toJSON(meta: any) {
+    // @TODO: Implement
+
+    return super.toJSON(meta)
+  }
+
+  dispose() {
+    // Assuming the geometry is not shared with other meshes
+    this.geometry.dispose()
+
+    this._matricesTexture?.dispose()
+    this._matricesTexture = null
+
+    return this
+  }
+
+  // @ts-ignore
+  onBeforeRender() {
+    if (this.material.defines) {
+      this.material.defines.BATCHING = true
+    }
+
+    // @TODO: Implement frustum culling for each geometry
+  }
+
+  // @ts-ignore
+  onAfterRender() {
+    if (this.material.defines) {
+      this.material.defines.BATCHING = false
+    }
+  }
+}
+
+export function convertToBatchedMesh(meshes: Mesh<BufferGeometry, Material>[]): BatchedMesh {
+  let totalVertices = 0
+  let totalIndices = 0
+
+  for (const mesh of meshes) {
+    totalVertices += mesh.geometry.attributes.position.count
+    totalIndices += mesh.geometry.index!.count
+  }
+
+  const material = meshes[0].material
+  const result = new BatchedMesh(meshes.length, totalVertices, totalIndices, material)
+
+  for (const mesh of meshes) {
+    const geoId = result.applyGeometry(mesh.geometry)
+    result.setMatrixAt(geoId, mesh.matrixWorld)
+  }
+
+  return result
+}
```

Comment:
Minor cleanup to drop ES6 methods. Should probably upstream and clarify how indexed versus non-indexed geometry should be handled.